mautic/mautic
# An optimistic locking mechanism for campaigns

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [❌]
| New feature/enhancement? (use the a.x branch)      | [✅]
| Deprecations?                          | [❌]
| BC breaks? (use the c.x branch)        | [❌]
| Automated tests included?              | [✅] 
| Related user documentation PR URL      |  
| Related developer documentation PR URL |  
| Issue(s) addressed                     | 



#### Description:

This PR adds support for optimistic locking. I tried to implement it via Doctrine functionality. It turned out to be too dangerous as it could cause many bugs (connected with OptimisticLockException). I implemented the optimistic locking the way we can control when an entity version is incremented/checked.

Let's consider the following steps:

1. Navigate to the edit page of an existing campaign.
2. Open a new  browser tab and navigate to the edit page of the same campaign.
3. In the latter tab modify the name (append `-2` for example) and click _Apply_.
4. Switch back to the first tab and without refreshing the browser modify the name (append `-3` for example) and click _Apply_.
5. **The modification from the third step is lost.**

#### Steps to test this PR:

1. Run the migration with the command `ddev exec bin/console doctrine:migrations:migrate` if needed
2. Navigate to the edit page of an existing campaign.
3. Open a new tab and navigate to the edit page of the same campaign.
4. In the latter tab modify the name (append `-2` for example) and click _Apply_.
5. Switch back to the first tab  and without refreshing the browser modify the name (append `-3` for example) and click _Apply_.
6. You should get an error message that reads `The record you are updating has been changed by someone else in the meantime. Please refresh the browser window and re-submit your changes.`.
7. The modification from the third step is not lost.
